nodedb.go

```diff
@@ -406,60 +407,124 @@ func (ndb *nodeDB) deleteLegacyNodes(version int64, nk []byte) error {
 	return ndb.batch.Delete(ndb.legacyNodeKey(nk))
 }
 
+var (
+	isDeletingLegacyVersionsMutex = &sync.Mutex{}
+	isDeletingLegacyVersions      = false
+)
+
 // deleteLegacyVersions deletes all legacy versions from disk.
 func (ndb *nodeDB) deleteLegacyVersions() error {
-	// Check if we have a legacy version
-	itr, err := ndb.getPrefixIterator(legacyRootKeyFormat.Key())
-	if err != nil {
-		return err
+	isDeletingLegacyVersionsMutex.Lock()
+	if isDeletingLegacyVersions {
+		isDeletingLegacyVersionsMutex.Unlock()
+		return nil
 	}
-	defer itr.Close()
+	isDeletingLegacyVersions = true
+	isDeletingLegacyVersionsMutex.Unlock()
 
-	// Delete orphans for all legacy versions
-	var prevVersion, curVersion int64
-	var rootKeys [][]byte
-	for ; itr.Valid(); itr.Next() {
-		legacyRootKeyFormat.Scan(itr.Key(), &curVersion)
-		rootKeys = append(rootKeys, itr.Key())
-		if prevVersion > 0 {
-			if err := ndb.traverseOrphans(prevVersion, curVersion, func(orphan *Node) error {
+	go func() {
+		defer func() {
+			isDeletingLegacyVersionsMutex.Lock()
+			isDeletingLegacyVersions = false
+			isDeletingLegacyVersionsMutex.Unlock()
+		}()
+
+		// Check if we have a legacy version
+		itr, err := ndb.getPrefixIterator(legacyRootKeyFormat.Key())
+		if err != nil {
+			ndb.logger.Error(err.Error())
+			return
+		}
+		defer itr.Close()
+
+		// Delete orphans for all legacy versions
+		var prevVersion, curVersion int64
+		var rootKeys [][]byte
+		counter := 0
+		for ; itr.Valid(); itr.Next() {
+			legacyRootKeyFormat.Scan(itr.Key(), &curVersion)
+			rootKeys = append(rootKeys, itr.Key())
+			if prevVersion > 0 {
+				if err := ndb.traverseOrphans(prevVersion, curVersion, func(orphan *Node) error {
+					counter++
+					if counter == 1000 {
+						counter = 0
+						time.Sleep(1000 * time.Millisecond)
```

Comment:
Is this sleep to create a write gap so that `SaveVersion` doesn't block too frequently from the newly introduced mutex in `BatchWithFlusher` while a long prune operation is ongoing?  

If `deleteLegacyVersions` is only called once on migration I guess it's OK, but this sleep is hard to understand and could result in unexpected wait times from the main I/o thread.  I'm also curious about the overhead of mutex/set release on every `Set/Delete` call.